File: packages/typescript/shared/src/models/embedding-graph.ts
export interface ChainEntity {
    id: string;
    ip?: string;
}

export interface VNFEntity extends ChainEntity {
    name?: string;
}

export interface ForwardingLink {
    source: ChainEntity;
    destination: ChainEntity;
    links: string[];
}

export interface VNF {
    host: ChainEntity;
    vnf: VNFEntity;
    next: VNF | VNF[] | string | string[];
    isTraversed?: boolean;
}

export type VNFUpdated = Pick<Partial<VNF>, "next"> & Omit<VNF, "next">;

export interface EmbeddingGraph {
    sfcID: string;
    vnfs: VNF;
    links: ForwardingLink[];
    isTraversed?: boolean;
}

export type EmbeddingGraphs = EmbeddingGraph[];

File: packages/python/shared/models/embedding_graph.py
"""
Defines the `EmbeddingGraphs` dictionary type and its associated types.
"""

from typing import Optional


class ChainEntity(dict):
    """
    Defines the `ChainEntity` dictionary type.
    """

    id: str
    ip: Optional[str]

class VNFEntity(ChainEntity):
    """
    Defines the `VNFEntity` dictionary type.
    """

    name: Optional[str]

class ForwardingLink:
    """
    Defines the `ForwardingLink` dictionary type.
    """

    source: ChainEntity
    destination: ChainEntity
    links: "list[str]"
    divisor: int


class VNF(dict):
    """
    Defines the `VNF` dictionary type.
    """

    host: ChainEntity
    vnf: VNFEntity
    next: "VNF | list[VNF] | str | list[str]"
    isTraversed: Optional[bool]


class EmbeddingGraph(dict):
    """
    Defines the `EmbeddingGraph` dictionary type.
    """

    sfcID: str
    vnfs: VNF
    links: "list[ForwardingLink]"
    isTraversed: Optional[bool]

# pylint: disable=invalid-name
EmbeddingGraphs = "list[EmbeddingGraph]"
